File: app/src/main/java/br/com/jonathanzanella/myexpenses/bill/EditBillActivity.java
package br.com.jonathanzanella.myexpenses.bill;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.EditText;

import org.apache.commons.lang3.StringUtils;
import org.joda.time.DateTime;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl;
import br.com.jonathanzanella.myexpenses.expense.Expense;
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyTextWatch;
import br.com.jonathanzanella.myexpenses.log.Log;
import br.com.jonathanzanella.myexpenses.validations.ValidationError;
import br.com.jonathanzanella.myexpenses.views.BaseActivity;
import butterknife.Bind;
import butterknife.OnClick;

public class EditBillActivity extends BaseActivity implements BillContract.EditView {
	public static final String KEY_BILL_UUID = "KeyBillUuid";

	@Bind(R.id.act_edit_bill_name)
	EditText editName;
	@Bind(R.id.act_edit_bill_amount)
	EditText editAmount;
	@Bind(R.id.act_edit_bill_due_date)
	EditText editDueDate;
	@Bind(R.id.act_edit_bill_init_date)
	EditText editInitDate;
	@Bind(R.id.act_edit_bill_end_date)
	EditText editEndDate;

	private final BillPresenter presenter;

	public EditBillActivity() {
		ExpenseRepository expenseRepository = new ExpenseRepository(new RepositoryImpl<Expense>(this));
		presenter = new BillPresenter(new BillRepository(new RepositoryImpl<Bill>(this), expenseRepository));
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_edit_bill);
	}

	@Override
	protected void onPostCreate(Bundle savedInstanceState) {
		super.onPostCreate(savedInstanceState);
		editAmount.addTextChangedListener(new CurrencyTextWatch(editAmount));
		presenter.onViewUpdated(false);
	}

	@Override
	protected void storeBundle(final Bundle extras) {
		super.storeBundle(extras);

		if(extras != null && extras.containsKey(KEY_BILL_UUID)) {
			new AsyncTask<Void, Void, Void>() {

				@Override
				protected Void doInBackground(Void... voids) {
					presenter.loadBill(extras.getString(KEY_BILL_UUID));
					return null;
				}

				@Override
				protected void onPostExecute(Void aVoid) {
					super.onPostExecute(aVoid);
					presenter.onViewUpdated(false);
				}
			}.execute();
		}
	}

	@Override
	protected void onSaveInstanceState(Bundle outState) {
		super.onSaveInstanceState(outState);
		String uuid = presenter.getUuid();
		if(uuid != null)
			outState.putString(KEY_BILL_UUID, uuid);
	}

	@Override
	protected void onStart() {
		super.onStart();
		presenter.attachView(this);
	}

	@Override
	protected void onStop() {
		super.onStop();
		presenter.detachView();
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.save, menu);
		return true;
	}

	@OnClick(R.id.act_edit_bill_init_date)
	void onInitDate() {
		presenter.onInitDate(this);
	}

	@Override
	public void onInitDateChanged(DateTime date) {
		editInitDate.setText(Bill.SIMPLE_DATE_FORMAT.format(date.toDate()));
	}

	@OnClick(R.id.act_edit_bill_end_date)
	void onEndDate() {
		presenter.onEndDate(this);
	}

	@Override
	public void onEndDateChanged(DateTime date) {
		editEndDate.setText(Bill.SIMPLE_DATE_FORMAT.format(date.toDate()));
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.action_save:
				presenter.save();
				break;
		}
		return super.onOptionsItemSelected(item);
	}

	@Override
	public void showBill(Bill bill) {
		editName.setText(bill.getName());
		editAmount.setText(CurrencyHelper.format(bill.getAmount()));
		editDueDate.setText(String.valueOf(bill.getDueDate()));
	}

	@Override
	public Bill fillBill(Bill bill) {
		String amountText = editAmount.getText().toString().replaceAll("[^\\d]", "");
		String dueDateText = editDueDate.getText().toString().replaceAll("[^\\d]", "");

		bill.setName(editName.getText().toString());
		bill.setAmount(StringUtils.isEmpty(amountText) ? 0 : Integer.parseInt(amountText));
		bill.setDueDate(StringUtils.isEmpty(dueDateText) ? 0 : Integer.parseInt(dueDateText));

		return bill;
	}

	@Override
	public void finishView() {
		Intent i = new Intent();
		i.putExtra(KEY_BILL_UUID, presenter.getUuid());
		setResult(RESULT_OK, i);
		finish();
	}

	@Override
	public void showError(ValidationError error) {
		switch (error) {
			case NAME:
				editName.setError(getString(error.getMessage()));
				break;
			case AMOUNT:
				editAmount.setError(getString(error.getMessage()));
				break;
			case DUE_DATE:
				editDueDate.setError(getString(error.getMessage()));
				break;
			case INIT_DATE:
			case INIT_DATE_GREATER_THAN_END_DATE:
				editInitDate.setError(getString(error.getMessage()));
				break;
			case END_DATE:
				editEndDate.setError(getString(error.getMessage()));
				break;
			default:
				Log.error(this.getClass().getName(), "Validation unrecognized, field:" + error);
		}
	}
}

File: app/src/main/kotlin/br/com/jonathanzanella/myexpenses/bill/EditBillActivity.kt
package br.com.jonathanzanella.myexpenses.bill

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.widget.EditText
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository
import br.com.jonathanzanella.myexpenses.helpers.CurrencyTextWatch
import br.com.jonathanzanella.myexpenses.helpers.toCurrencyFormatted
import br.com.jonathanzanella.myexpenses.transaction.Transaction
import br.com.jonathanzanella.myexpenses.validations.ValidationError
import br.com.jonathanzanella.myexpenses.views.anko.*
import org.apache.commons.lang3.StringUtils
import org.jetbrains.anko.*
import org.joda.time.DateTime
import timber.log.Timber

class EditBillActivity : AppCompatActivity(), BillContract.EditView {
    override val context = this
    private val presenter: BillPresenter
    private val ui = EditBillActivityUi()

    init {
        val expenseRepository = ExpenseRepository()
        presenter = BillPresenter(BillRepository(expenseRepository))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ui.setContentView(this)

        storeBundle(savedInstanceState)
        storeBundle(intent.extras)

        ui.toolbar.setup(this)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        ui.editAmount.addTextChangedListener(CurrencyTextWatch(ui.editAmount))
        ui.editInitDate.onClick { presenter.onInitDate(this) }
        ui.editEndDate.onClick { presenter.onEndDate(this) }
        presenter.onViewUpdated(false)
    }

    fun storeBundle(extras: Bundle?) {
        if (extras?.containsKey(KEY_BILL_UUID) == true) {
            doAsync {
                presenter.loadBill(extras.getString(KEY_BILL_UUID))

                uiThread { presenter.onViewUpdated(false) }
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        val uuid = presenter.uuid
        if (uuid != null)
            outState.putString(KEY_BILL_UUID, uuid)
    }

    override fun onStart() {
        super.onStart()
        presenter.attachView(this)
    }

    override fun onStop() {
        super.onStop()
        presenter.detachView()
    }

    override fun setTitle(string: String) {
        ui.toolbar.title = string
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.save, menu)
        return true
    }

    override fun onInitDateChanged(date: DateTime) {
        ui.editInitDate.setText(Transaction.SIMPLE_DATE_FORMAT.format(date.toDate()))
    }

    override fun onEndDateChanged(date: DateTime) {
        ui.editEndDate.setText(Transaction.SIMPLE_DATE_FORMAT.format(date.toDate()))
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_save -> presenter.save()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun showBill(bill: Bill) {
        ui.editName.setText(bill.name)
        ui.editAmount.setText(bill.amount.toCurrencyFormatted())
        ui.editDueDate.setText(bill.dueDate.toString())
    }

    override fun fillBill(bill: Bill): Bill {
        val amountText = ui.editAmount.text.toString().replace("[^\\d]".toRegex(), "")
        val dueDateText = ui.editDueDate.text.toString().replace("[^\\d]".toRegex(), "")

        bill.name = ui.editName.text.toString()
        bill.amount = if (StringUtils.isEmpty(amountText)) 0 else Integer.parseInt(amountText)
        bill.dueDate = if (StringUtils.isEmpty(dueDateText)) 0 else Integer.parseInt(dueDateText)

        return bill
    }

    override fun finishView() {
        val i = Intent()
        i.putExtra(KEY_BILL_UUID, presenter.uuid)
        setResult(Activity.RESULT_OK, i)
        finish()
    }

    override fun showError(error: ValidationError) {
        when (error) {
            ValidationError.NAME -> ui.editName.error = getString(error.message)
            ValidationError.AMOUNT -> ui.editAmount.error = getString(error.message)
            ValidationError.DUE_DATE -> ui.editDueDate.error = getString(error.message)
            ValidationError.INIT_DATE, ValidationError.INIT_DATE_GREATER_THAN_END_DATE -> ui.editInitDate.error = getString(error.message)
            ValidationError.END_DATE -> ui.editEndDate.error = getString(error.message)
            else -> Timber.e("Validation unrecognized, field:" + error)
        }
    }

    companion object {
        val KEY_BILL_UUID = "KeyBillUuid"
    }
}

class EditBillActivityUi : AnkoComponent<EditBillActivity> {
    lateinit var toolbar : TemplateToolbar
    lateinit var editName: EditText
    lateinit var editAmount: EditText
    lateinit var editDueDate: EditText
    lateinit var editInitDate: EditText
    lateinit var editEndDate: EditText

    override fun createView(ui: AnkoContext<EditBillActivity>) = with(ui) {
        verticalLayout {
            toolbar = toolbarTemplate {}

            linearViewFrame {
                textInputLayout {
                    editName = appCompatEditText {
                        id = R.id.act_edit_bill_name
                        hint = resources.getString(R.string.name)
                    }
                }
                textInputLayout {
                    editAmount = appCompatEditText {
                        id = R.id.act_edit_bill_amount
                        hint = resources.getString(R.string.amount)
                    }
                }
                textInputLayout {
                    editDueDate = appCompatEditText {
                        id = R.id.act_edit_bill_due_date
                        hint = resources.getString(R.string.due_date)
                    }
                }
                textInputLayout {
                    editInitDate = clickableView {
                        id = R.id.act_edit_bill_init_date
                        hint = resources.getString(R.string.start)
                    }
                }
                textInputLayout {
                    editEndDate = clickableView {
                        id = R.id.act_edit_bill_end_date
                        hint = resources.getString(R.string.end)
                    }
                }
            }
        }.applyRecursively(::applyTemplateViewStyles)
    }
}
